[WIP] Tensor decoder for image segmentation

L2zz opened the pull request:
Tensor decoder for image segmentation.

Based on tensordec-directvideo.c and tensordec-boundingbox.c.
Implement skeleton code of tensordec-imagesegment.c
Add meson build for it.

Signed-off-by: niklasjang niklasjang@gmail.com
Signed-off-by: Jihoon Lee ulla4571@gmail.com

kparichay committed:
[tests] Install models/data with tests

Install models/data with tests correctly when needed
Install tizen-capi-unittest with install-test option

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

jaeyun-jung committed:
[Security] handle returned value

Check mem allocation failure case and refactor file-close in IIO.

This PR is related with ssms-1016-0006.

Signed-off-by: Jaeyun Jung <jy1210.jung@samsung.com>

mojunsang26 committed:
[CodeClean] combine if statements
combine if statements in tensor_transform.c.
I confirmed it worked well in example_image_classification_tflite.

Signed-off-by: Junsang Mo <junsang.mo@samsung.com>

leemgs committed:
Added the coverity badge

This commit is to append the badge of the Coverity (static analysis tool).
TODO: the Coverity module of TAOS-CI will be soon enabled in the nnstreamer GitHub repository.

Signed-off-by: Geunsik Lim <geunsik.lim@samsung.com>

Lee-WonJun committed:
Add execution bit of gradlew
Added excution bit of gradlew to prevent permission denied error.

Signed-off-by: 이원준 <dldnjs1013@nate.com>

abcinje committed:
[Typo] Fix typo

This patch fixes typo in packaging/nnstreamer.spec

Signed-off-by: Injae Kang <abcinje@gmail.com>

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,350 @@
+  guint **segment_map;  /**< The image segmentated map */
+
+  /*  From option2 */
+  guint o_width; /**< Output video Width */
```
NOTE THAT THIS OPTION IS ONLY FOR TFLITE_DEEPLAB_IMAGE_SEGMENT!

L2zz replied:
Sry, I don't know what to do for it.

L2zz replied:
Does it mean do not place the option dependent value in common data structure?

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,350 @@
+ */
+typedef enum
+{
+  TFLITE_DEEPLAB_IMAGE_SEGMENT = 0,
```
SEGMENT? SEGMENTATION? 

Actually, this decoder is for image segment(ation), then why do we need the word, image_segment here and there?

L2zz replied:
I feel it is too long to use it. If it is hard to understand, I'll change it.

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,350 @@
+  idata->mode = UNKNOWN_IMAGE_SEGMENT;
+  idata->i_width = 0;
+  idata->i_height = 0;
+  idata->o_width = 0;
```
Do we really need o_width and o_height?

L2zz replied:
May be not. I'll remove it

L2zz replied:
Do we need to get i_width and i_height by option?

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,350 @@
+      GST_ERROR
+          ("mode-option-2 of image segmentation is video output dimension (WIDTH:HEIGHT). "
+          "The given parameter, \"%s\", is not acceptable.", param);
+      return TRUE;              /* Ignore this param */
```
TRUE?

L2zz replied:
TRUE means just ignore wrong option and keep going pipeline.

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,350 @@
+      GST_ERROR
+          ("mode-option-3 of image segmentation is input video dimension (WIDTH:HEIGHT). "
+          "The given parameter, \"%s\", is not acceptable.", param);
+      return TRUE;              /* Ignore this param */
```
ditto

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,350 @@
+    const GstTensorMemory * input, GstBuffer * outbuf)
+{
+  image_segments *idata = *pdata;
+  const size_t size = idata->i_width * idata->i_height * 4;
```
eliminate hard-coded 4. use pre-defined one.

L2zz replied:
Yes, I'll change it

wooksong reviewed (changes requested):
Need to improve

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,287 @@
+  image_segments *idata = *pdata;
+
+  if (op_num == 0) {
+    /* option1 = image segmentation decoder mode */
```
option == 0 and option1 keep confusing me.
```
The first option indicates a mode of this decoder sub-plugin.
```

Or you use your own phrase. Anyway change it.

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,287 @@
+void init_is (void) __attribute__ ((constructor));
+void fini_is (void) __attribute__ ((destructor));
+
+#define RGBA_CHANNEL                   4
```
Is this only work for RGBA?

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,287 @@
+void fini_is (void) __attribute__ ((destructor));
+
+#define RGBA_CHANNEL                   4
+#define TFLITE_DEEPLAB_TOTAL_LABELS    21
```
ditto

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,287 @@
+typedef enum
+{
+  TFLITE_DEEPLAB = 0,
+  UNKNOWN,
```
MODE_TFILITE_DEEPLAB
MODE_UNKNOWN

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,287 @@
+    return TRUE;
+  }
+
+  GST_WARNING ("mode-option-\"%d\" is not definded.", op_num);
```
mode-option or just mode?

L2zz replied:
Mode-option, if user write invalid option like option3(we don't have), then warn that case.

wooksong commented on ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c:
```diff
@@ -0,0 +1,287 @@
+  if (gst_buffer_get_size (outbuf) == 0)
+    gst_buffer_append_memory (outbuf, out_mem);
+
+  return GST_FLOW_OK;
```
Q: Does this function always return GST_FLOW_OK unless it dies with assertion?

L2zz replied:
Yes

TunaLee committed:
[CodeClean] replace meaningless variables
rename a and b to other more meaningful ones.
a -> from, b -> to

Signed-off-by: Dongwon Lee <dongwon1103@naver.com>

L2zz committed:
[WIP] Tensor decoder for image segmentation

In order to implement tensor decoder for image segmentation,
this patch composes skeleton code for it.

Signed-off-by: niklasjang <niklasjang@gmail.com>
Signed-off-by: Jihoon Lee <ulla4571@gmail.com>

L2zz committed:
[Decoder/Meson] Meson build for image segmentation decoder

In order to add tensor decoder for image segmentation to build tree,
updates meson build script.

Signed-off-by: niklasjang <niklasjang@gmail.com>
Signed-off-by: Jihoon Lee <ulla4571@gmail.com>
